### PR TITLE
feat: configurable ls path [ROAD-1082]

### PR DESCRIPTION
**Pull request by j-luong**
### Description

_This PR should enable a setting in the Snyk settings menu to provide a custom Language Server binary_

### Checklist

- [x] Tests added and all succeed
- [x] Linted
- [x] CHANGELOG.md updated

### CODE REVIEW

**Comment by j-luong** [comment]
Currently manual testing shows it working, but when changing the LS binary path, the old Language Server output window doesn't close, leading to a buildup of Language Server output windows:
Screenshot 2022-09-08 at 16 32 06

File: src/snyk/common/languageServer/lsExecutable.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import path from 'path';
+import { Checksum } from '../../cli/checksum';
+import { SupportedPlatform } from '../../cli/supportedPlatforms';
+import { Platform } from '../../common/platform';
+
+export class LsExecutable {
+  // If values updated, `.vscodeignore` to be changed.
+  public static filenameSuffixes: Record<SupportedPlatform, string> = {
+    linux: 'snyk-ls-linux',
```
Comment (bastiandoetsch) [inline src/snyk/common/languageServer/lsExecutable.ts:9]: It's a bit more complicated... please have a look here: https://github.com/snyk/vscode-extension/compare/main...feat/ROAD-1054_manage_ls_binary_automatically#diff-3780f20a3e492c878059b07f3512993edde0e01b86fc3a2f9f4b5fba8bf821f1R1

**Comment by j-luong** [comment]
Removed feature `SCENARIO LS restart when changing custom LS`.

There are snyc issues caused by the need to synchronise the LanguageClient in instantiation. A manual VSCode restart will be required when changing the LS binary path.

File: src/test/unit/common/configuration.test.ts [hunk 2]
```diff
@@ -198,17 +198,4 @@ suite('Configuration', () => {
 
     deepStrictEqual(configuration.getPreviewFeatures(), previewFeatures);
   });
-
```
Comment (michelkaporin) [inline src/test/unit/common/configuration.test.ts:201]: Shouldn't we still test that logic with setting value?

File: src/snyk/common/languageServer/lsExecutable.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  }
+
+  static getPath(extensionDir: string, customPath?: string): string {
+    if (customPath) {
+      return customPath;
+    }
+
+    const platform = Platform.getCurrentWithArch();
+    const fileName = LsExecutable.getFilename(platform);
+    const lsPath = path.join(extensionDir, fileName);
```
Comment (michelkaporin) [inline src/snyk/common/languageServer/lsExecutable.ts:31]: This is wrong, as the language server will be stored inside of XDG dir to reuse across different IDEs. See 5th scenario in the ticket description.

**Review by michelkaporin — CHANGES_REQUESTED** [review]

File: package.json [hunk 4]
```diff
@@ -331,7 +336,7 @@
           "when": "!snyk:loggedIn && snyk:previewLsAuthentication"
         },
         {
-          "command": "snyk.logout",
+          "command": "snyk.logout.old",
```
[the remark this answers is not in the document]
Comment (j-luong) [reply package.json]: It was added by mistake, originally it was so I could manual test it. Will remove.

File: src/snyk/common/platform.ts [hunk 5]
```diff
@@ -8,4 +9,24 @@ export class Platform {
   static getVersion(): string {
     return `${os.release()}-${os.arch}`;
   }
+
+  static getHomeDir(): string {
```
[the remark this answers is not in the document]
Comment (j-luong) [reply src/snyk/common/platform.ts]: It isn't, will remove.

File: src/snyk/common/languageServer/lsExecutable.ts (hunk 3 above)
Comment (j-luong) [reply src/snyk/common/languageServer/lsExecutable.ts:30]: This was to keep it consistent with the CLI implementation, but also, do we want to centralise LS like this? Could having separate IDEs manage the same binary lead to unexpected issues? e.g. what would happen is an IDE is already open and a separate IDE gets opened and updates LS?

File: .vscodeignore [hunk 6]
```diff
@@ -29,3 +29,11 @@ snyk-macos
 snyk-linux
 snyk-win.exe
 
+# Snyk LS
```
[the remark this answers is not in the document]
Comment (j-luong) [reply .vscodeignore]: will remove

File: src/snyk/common/languageServer/lsExecutable.ts (hunk 3 above)
Comment (michelkaporin) [reply src/snyk/common/languageServer/lsExecutable.ts:30]: That was the decision we made when starting with LS integration in Eclipse. You can cross verify by @bastiandoetsch. Your points are totally valid though, maybe it's not that likely that both updates start 100% simultaneously. And update generally should create a lock file, that would prevent other IDEs from starting their own update.

Comment (j-luong) [reply src/snyk/common/languageServer/lsExecutable.ts:30]: Happy to revert it if that's what we want, this change was raised during the last sprint after seeing how it was implemented for the CLI; and all seemed ok with it.

File: src/snyk/common/configuration/configuration.ts [hunk 7]
```diff
@@ -184,9 +188,62 @@ export class Configuration implements IConfiguration {
...
-    if (!this.processEnv.SNYK_LS_PATH) throw new Error('SNYK_LS_PATH environment variable is not set.');
-    return this.processEnv.SNYK_LS_PATH;
+  getSnykLanguageServerPath(): string | undefined {
+    return this.workspace.getConfiguration<string>(
+      CONFIGURATION_IDENTIFIER,
+      this.getConfigName(ADVANCED_CUSTOM_LS_PATH),
+    );
+  }
+
+  getSnykLanguageServerPath2(): string {
```
Comment (michelkaporin) [inline src/snyk/common/configuration/configuration.ts:198]: That doesn't belong to `configuration`. It's LS-specific logic.

File: src/snyk/common/configuration/configuration.ts [hunk 8]
```diff
@@ -184,9 +188,62 @@ export class Configuration implements IConfiguration {
...
+    // set default home directory
+    const homeDir = Platform.getHomeDir();
+
+    const lsPlatform = LsExecutable.getCurrentWithArch();
+
+    // set Language Server filename
+    const lsFilename = LsExecutable.filenameSuffixes[lsPlatform];
+
+    // set default Language Server path
+    const defaultPaths: Record<LsSupportedPlatform, string> = {
```
Comment (michelkaporin) [inline src/snyk/common/configuration/configuration.ts:214]: Let's align with Eclipse implementation https://github.com/snyk/snyk-eclipse-plugin/blob/2d455fc827c7ba203b8999f358b62a0ae021a979/plugin/src/main/java/io/snyk/eclipse/plugin/utils/FileSystemUtil.java#L34

General tip: avoid having such lists as they're hard to maintain when new os/arch support being added.

Comment (j-luong) [reply src/snyk/common/configuration/configuration.ts]: As suggested in Slack, I've aligned with the [Language Server implementation for the CLI](https://github.com/snyk/snyk-ls/blob/c7fad4a45c94bf76a23b05c0b0038f352155044f/application/config/config.go#L438).

It uses the [XDG package](https://pkg.go.dev/github.com/adrg/xdg@v0.4.0#section-readme).

**Review by bastiandoetsch — APPROVED** [review]
LGTM

**Pull request merged**